File: Algorithm/Python/973. K Closest Points to Origin.py
# -*- coding: utf-8 -*-
# Created on Mon Jan 14 2019 18:56:19
# Author: WuLC
# EMail: liangchaowu5@gmail.com

# heap queue 

import heapq

class Solution(object):
    def kClosest(self, points, K):
        """
        :type points: List[List[int]]
        :type K: int
        :rtype: List[List[int]]
        """
        result = []
        for p in points:
            if len(result) < K:
                heapq.heappush(result, (-(p[0]**2 + p[1]**2), p)) # can't use result.append(-(p[0]**2 + p[1]**2), p) in ordre to heapify the 
            else:
                heapq.heappushpop(result, (-(p[0]**2 + p[1]**2), p))
        return map(lambda x: x[1], result)

File: Algorithm/C++/973. K Closest Points to Origin.cc
/*
 * Created on Mon Jan 14 2019 20:28:8
 * Author: WuLC
 * EMail: liangchaowu5@gmail.com
 */
// priority queue with pair
#include <vector>
#include <queue>
#include <utility>

using std::vector;
using std::priority_queue;

class Solution {
 public:
  vector<vector<int>> kClosest(vector<vector<int>>& points, int K) {
    priority_queue<std::pair<int, vector<int>>> queue; // default max heap
    for (auto p : points) {
      queue.push(std::make_pair(p[0]*p[0] + p[1]*p[1], p));
      if (queue.size() > K) queue.pop();
    }
    vector<vector<int>> result;
    while (!queue.empty()) {
      result.push_back(queue.top().second);
      queue.pop();
    }
    return result;
  }
};